Extract and optimize group member calculation

tnqn opened the pull request:
Previously grouping member calculation was coupled with the NetworkPolicyController, and there were two kinds of heavy calculation in the process:
For each event like Pod add, it needed to find all affected groups by matching the new Pod's labels with all groups that can potentially select it. If it was an update event, it even needed to perform the process one more time for its old labels. Then it triggered the sync process of the affected groups.
For the sync process of a group, it needed to find all Pods or ExternalEntities by matching its selectors with all entities that can potentially match it.
For each pair of Pod and Group, they would be matched at least twice. Besides, whenever the group was processed another time, it would again scan all Pods.

The repeated calculation can be eliminated by caching the result of the first matching process when processing Pod events. Then the matching process will only occur when there's a new Pod or a new Group.
Besides, most Pods actually share same labels with others when they are managed by same controller. The cache can be further optimized to maintain the relationship between label set and label selector, making Pods that having same labels and Groups that having same selectors share the cache.

This patch introduces an interface which is responsible for group member calculation and an implementation implementing it with the above optimization. The interface is then consumed by NetworkPolicyController and features like EndpointQuerier and ClusterGroup to retrieve Pods/ExternalEntities selected by a given Group or Groups that select a given Pod. Besides, other controllers that have the group logic like EgressPolicy can consume it directly without having redundant code with NetworkPolicyController.

Performance impact:
```
name                                       old time/op    new time/op    delta
InitXLargeScaleWithSmallNamespaces-48         3.65s ±14%     4.35s ± 7%  +19.20%  (p=0.016 n=5+5)
InitXLargeScaleWithLargeNamespaces-48         5.34s ± 2%     1.48s ± 5%  -72.34%  (p=0.008 n=5+5)
InitXLargeScaleWithOneNamespace-48            3.22s ± 8%     3.46s ± 2%     ~     (p=0.056 n=5+5)
InitXLargeScaleWithNetpolPerPod-48            45.9s ±11%     25.6s ± 4%  -44.18%  (p=0.008 n=5+5)
InitXLargeScaleWithClusterScopedNetpol-48     1.40s ± 3%     1.18s ± 5%  -16.08%  (p=0.008 n=5+5)

name                                       old alloc/op   new alloc/op   delta
InitXLargeScaleWithSmallNamespaces-48         824MB ± 0%     900MB ± 0%   +9.22%  (p=0.008 n=5+5)
InitXLargeScaleWithLargeNamespaces-48         629MB ± 0%     321MB ± 0%  -49.00%  (p=0.008 n=5+5)
InitXLargeScaleWithOneNamespace-48            921MB ± 0%     922MB ± 0%   +0.13%  (p=0.008 n=5+5)
InitXLargeScaleWithNetpolPerPod-48           3.39GB ± 0%    0.13GB ± 0%  -96.19%  (p=0.008 n=5+5)
InitXLargeScaleWithClusterScopedNetpol-48     279MB ± 0%     244MB ± 0%  -12.59%  (p=0.008 n=5+5)

name                                       old allocs/op  new allocs/op  delta
InitXLargeScaleWithSmallNamespaces-48         12.0M ± 0%     12.6M ± 0%   +5.35%  (p=0.008 n=5+5)
InitXLargeScaleWithLargeNamespaces-48         4.29M ± 0%     4.48M ± 0%   +4.63%  (p=0.008 n=5+5)
InitXLargeScaleWithOneNamespace-48            1.39M ± 0%     1.40M ± 0%   +0.74%  (p=0.008 n=5+5)
InitXLargeScaleWithNetpolPerPod-48            1.66M ± 0%     1.75M ± 0%   +5.52%  (p=0.008 n=5+5)
InitXLargeScaleWithClusterScopedNetpol-48     3.07M ± 0%     3.20M ± 0%   +4.04%  (p=0.008 n=5+5)
```

Explanation:
- `InitXLargeScaleWithSmallNamespaces` has only 4 Pods in each Namespace and all NetworkPolicies are namespace scoped, so even without the index, each policy just needs to scan 4 Pods at most. Maintaing the index adds few overhead, hence the minor increment in time and memory.
- `InitXLargeScaleWithLargeNamespaces` has 1000 Pods and 100 NetworkPolicies in each Namespace, each NetworkPolicy applies to 10 Pods. Pods that have same labels will only be scaned once so many calculation is saved. When syncing groups, the result can be got directly without listing all Pods in the Namespace first, so calculation and memory is saved.
- `InitXLargeScaleWithOneNamespace` just has many Pods but only 1 namespace, 1 AppliedToGroup and 1 AddressGroup in total, so having the index doesn't help anything.
- `InitXLargeScaleNetpolPerPod` has only 1 namespace, 10000 Pods and 1 NetworkPolicy per Pod. It can benefit from the index greatly for the same reason as `InitXLargeScaleWithLargeNamespaces`.

tnqn commented:
@jianjuns @antoninbas @Dyanngg Please let me know if the change makes sense to you. And whether it can help nested group feature or make it harder? @Dyanngg 
I will add performance impact later.

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,719 @@
+	// The keys of the groupItems that share the selectorItem.
+	groups sets.String
+	// The keys of the labelItems that match the selectorItem.
+	labels sets.String
```
could you rename this member to `labelKeys`?

tnqn replied:
Sure, will make labelItem and its keys more explicit.

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,719 @@
+
+// getLabelKey returns the label key used in labelItems.
+func getLabelKey(entityType entityType, obj metav1.Object) string {
+	return string(entityType) + "/" + obj.GetNamespace() + "/" + labels.Set(obj.GetLabels()).String()
```
`labels.Set(obj.GetLabels()).String()`: is the order of labels deterministic? Ultimately, if I define PodA with 2 labels in a certain order, then PodB with the same 2 labels in the opposite order, will the same string be returned. By order, I mean the order I use in a YAML declaration, or the order in which I insert labels in the map. And when I call `labels.Set(obj.GetLabels()).String()` multiple times on the same object, it's guaranteed to always return the same string?

tnqn replied:
Yes, labels.Set sorts it.
```
func (ls Set) String() string {
	selector := make([]string, 0, len(ls))
	for key, value := range ls {
		selector = append(selector, key+"="+value)
	}
	// Sort for determinism.
	sort.StringSlice(selector).Sort()
	return strings.Join(selector, ",")
}
```

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,719 @@
+}
+
+// getLabelKey returns the label key used in labelItems.
+func getLabelKey(entityType entityType, obj metav1.Object) string {
```
I find the term "label key" a bit confusing because in K8s a "label key" already means something (the "name" of the label).

tnqn replied:
updated to `getLabelItemKey`, so do others.

antoninbas commented on pkg/controller/grouping/group_entity_index_test.go:
```diff
@@ -0,0 +1,314 @@
+)
+
+var (
+	// Fake Pods
```
would it be worth having a test with different label names (not just "app")? Or do you think the coverage would be no different?

tnqn replied:
I guess the converage should be same. podBar1 and podFoo1InOtherNamespace are used as noise and to ensure pod label selector and namespace selector are respected. And we are using K8s lib's method to check whether a selector matches a label set, the labels' content is transparent to Antrea.

Dyanngg commented:
> @jianjuns @antoninbas @Dyanngg Please let me know if the change makes sense to you. And whether it can help nested group feature or make it harder? @Dyanngg
> I will add performance impact later.

I'm still reviewing the `GroupEntityIndex` interface, but after the first look I don't think it will make nested group computation easier or more complicated. It's just that the nested group member calc logic would be entirely re-written, since right now I rely on the `groupMember` field of the internalGroup to union nested groupMembers, which is removed by this PR.

antoninbas commented on pkg/controller/networkpolicy/networkpolicy_controller_perf_test.go:
```diff
@@ -127,6 +127,58 @@ func TestInitXLargeScaleWithSmallNamespaces(t *testing.T) {
 	testComputeNetworkPolicy(t, 10*time.Second, namespaces, networkPolicies, pods)
 }
 
+func TestInitXLargeScaleWithLargeNamespaces(t *testing.T) {
```
Is there a big improvement (computation time) for this test case? I remember when I was experimenting with DDlog (https://drive.google.com/file/d/1V1fMQ7VQaFPp-AO1aXbYC-fPoKPwo0gC/view), there was one single-namespace test case (TestPerf1) for which the Antrea Controller performed very badly.

tnqn replied:
I'm still working on verifying the performance impact. I found the existing two cases are not really improved and suspected it's was because in TestInitXLargeScaleWithSmallNamespaces there were only a few pods in each namespace and it has only namespace scoped group so number of potential pod candidates (4 pods) that will be processed for each group doesn't make a difference, while in TestInitXLargeScaleWithOneNamespace, there was actually only single group, and the group is typically only handled less than 10 times, so having a cache didn't make a difference too.
I added this case which should be closer to practice that there is a good number of Pods and different groups in each Namespace to verify how much the cache can help. The result is promising: the execution time was reduced from 14 ~ 15s to 5 ~ 6s. I can check TestPerf1 too.

antoninbas reviewed (commented):
I am fine with the approach. What's the memory usage impact of the change when running the benchmarks in networkpolicy_controller_perf_test.go?

tnqn commented:
> > @jianjuns @antoninbas @Dyanngg Please let me know if the change makes sense to you. And whether it can help nested group feature or make it harder? @Dyanngg
> > I will add performance impact later.
> 
> I'm still reviewing the `GroupEntityIndex` interface, but after the first look I don't think it will make nested group computation easier or more complicated. It's just that the nested group member calc logic would be entirely re-written, since right now I rely on the `groupMember` field of the internalGroup to union nested groupMembers, which is removed by this PR.

I removed the field as it was consumed by `GetGroupMembers`, `populateAddressGroupMemberSet` and `getAppliedToWorkloads` when they need to know the members of a clustergroup but with the change it can be achieved by querying the `GroupEntityIndex` directly, so the internalGroup itself doesn't have to store the redundant data. I can imagine that in nested group case, you want to avoid calculate the union of the members again when querying its member, right? If I understand correctly, I can add this field back or let's see whether there is more efficient way.

tnqn commented on pkg/controller/grouping/controller.go:
```diff
@@ -0,0 +1,239 @@
+)
+
+const (
+	controllerName = "GroupController"
```
Sure, sounds better, will update.

jianjuns replied:
GroupEntityController?

tnqn commented on pkg/controller/grouping/controller.go:
```diff
@@ -0,0 +1,239 @@
+
+const (
+	controllerName = "GroupController"
+	// NetworkPolicyController is the only writer of the antrea network policy
```
thanks for reminding, will corrent all comments.

tnqn commented on pkg/controller/networkpolicy/clustergroup.go:
```diff
@@ -366,24 +321,30 @@ func (n *NetworkPolicyController) serviceToGroupSelector(service *v1.Service) *a
-	groups, err := n.internalGroupStore.GetByIndex(store.GroupMemberIndex, memberKey)
-	if err != nil {
-		return nil, fmt.Errorf("failed to retrieve Group associated with key %s", memberKey)
+	groups, _ := n.groupEntityIndex.GetGroupsForPod(namespace, name)
+	clusterGroups, exists := groups[clusterGroupType]
```
I should add a TODO here. I found the original code actually got an union of groups that a Pod having the provided name and an Entity having the provided name belongs to. Not sure whether it's by design.
@Dyanngg could you confirm? should I return both groups here?

tnqn replied:
Updated to try Pod first, then ExternalEntity

tnqn reviewed (commented):
@Dyanngg and @antoninbas thanks for your quick review and feedback.

tnqn commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,719 @@
+// that managed by a deployment controller have same labels.
+//
+// It maintains indexes between label set and label selector so that querying label sets that match a label selector and
+// reversed queries can be performed with a consistent time complexity. Indirectly, querying entities that match a group
```
I meant something like O(1) time complexity for this kind queries and reverse queries. Maybe I should just use O(1)?

antoninbas replied:
so you mean "constant" not "consistent" :)

tnqn replied:
right, thanks!

tnqn commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,719 @@
+// GroupEntityIndex provides methods to query entities that a given group selects and groups that select a given entity.
+// It maintains indexes between groups and entities to make the query efficient. It supports callers to register
+// callbacks that will be called when a specific type of groups' entities are updated.
+type GroupEntityIndex interface {
```
Yes, it can be used for that too.

jianjuns commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,719 @@
+type entityType string
+
+const (
+	podEntityType      entityType = "Pod"
```
If it is used internally only, can we use shorter strings like "p" or "e", hashing/comparison etc. can be faster?

jianjuns commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,719 @@
+	labelItems map[string]*labelItem
+	// labelIndex is nested map from entityType to Namespace to keys of labelItems.
+	// It's used to filter potential labelItems when matching a namespace scoped selectorItem.
+	labelIndex map[entityType]map[string]sets.String
```
Is it more efficient to create two separate maps for Pods and ExternalEntities?

jianjuns replied:
Or to share code, we can convert them to one internal type?

tnqn replied:
The code of matching Pods and ExtenalEntities is already shared and they are treated in same way when stored in labelItems.
Separating them in labelIndex is to reduce the potential labelItems a selector needs to match. This is because a selector can either select ExternalEntity or Pod. If it's a Pod selector, this is no need to try ExternalEntity labelItems. The same reason applies to selectorIndex.

jianjuns reviewed (commented):
I also wonder the memory usage.

And do you think any in-memory store implementation can help (compared to defining all maps case by case)?

tnqn commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,719 @@
+	}
+
+	i.namespaceLabels[namespace.Name] = namespace.Labels
+	// Resync cluster scoped selectors as they may start or stop matching the namespace because of the label update.
```
done

tnqn commented:
@jianjuns @antoninbas I updated benchmark test result in PR description. To get accurate cpu and memory metrics, I used benchmark test and execute the processing in serial manner.

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,728 @@
+	// GetEntities returns the selected Pods or ExternalEntities for the given group.
+	GetEntities(groupType string, name string) ([]*v1.Pod, []*v1alpha2.ExternalEntity)
+	// GetGroupsForPod returns the groups that select the given Pod.
+	GetGroupsForPod(namespace, name string) (map[string][]string, bool)
```
do you think we could define a new type for the group type: `type GroupType string`? It would help clarify what the returned values are if we have `map[GroupType][]string` instead of `map[string][]string`.

jianjuns replied:
Maybe an enum?

jianjuns replied:
I feel enum (int) also works, but I guess your concern is harder to track the int values when they are not in the same package?

antoninbas replied:
> Maybe I should define GroupType (using string) here and let consumers define their own values in their own packages, does it make sense to you?

That works for me as a middle ground between using a plain `string` and defining named constants for all group types in this package.

tnqn replied:
@jianjuns  Yes, if using enum(int), they we should define them together in this package to avoid conflict.

jianjuns replied:
I think it is just like string is easier to remember and has less chance to conflict (but can still). I do not like long string lookup, but I understand your point and no strong opinion.

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,728 @@
+		}
+		return true
+	}
+	// TODO: if none of the above selector is set, it means matching all entities or none?
+	// This is only possible for ClusterGroup. From previous code, it means matching none.
```
I think @Dyanngg should confirm. If one wants to select all Pods in all Namespaces, I assume we need an empty `PodSelector` and an empty `NamespaceSelector` (as opposed to `nil` ones)?

Dyanngg replied:
This is an undocumented case so I think we should keep it consistent with previous releases (matches nothing). As you can see such semantics is quite implicit. That's why I'm proposing to make `namespaceSelector` a required field for ClusterGroup and ACNP appliedTo for the next API version. This would eliminate such ambiguity.

tnqn replied:
The current code keeps previous behavior

Dyanngg replied:
yes, and I would agree with the current implementation.

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,728 @@
+	switch oldValue := oldEntity.(type) {
+	case *v1.Pod:
+		// For Pod, we only care about PodIP and NodeName update.
+		// Some other attributes we care are immutable, e.g. the named ContainerPort.
```
s/we care/we care about

tnqn replied:
done

antoninbas commented on pkg/controller/networkpolicy/networkpolicy_controller_perf_test.go:
```diff
@@ -18,6 +18,7 @@ package networkpolicy
 
```
perhaps unrelated to this PR, but I noticed a long time ago that we don't have any benchmark for policies using `NamespaceSelector` to select network peers. Could we add one?

tnqn replied:
will add one

tnqn replied:
Thanks for bring this up. I added a new case and found the new implementation made it worse for execution time. The reason was in previous code we got all namespaces that match namespaceSelector first then scan Pods in them. The new code scans all labels in all namespaces in this case so it became worse. I used the same way as previous code to optimize it, now it's better than previous code.

antoninbas reviewed (commented):
the changes look good to me, and so do the benchmark results

Dyanngg commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,728 @@
+
+// deletePodFromLabelItem disconnects an entityItem from a labelItem.
+// The labelItem will be deleted if it's no longer used by any entityItem.
+func (i *GroupEntityIndex) deletePodFromLabelItem(label, entity string) *labelItem {
```
I assume this function works for ExternalEntity as well?

Dyanngg reviewed (commented):
LGTM overall

jianjuns commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,728 @@
+func (i *GroupEntityIndex) notify(selector string) {
+	sItem := i.selectorItems[selector]
+	for group := range sItem.groupItemKeys {
+		i.eventChan <- group
```
This can be a block call? Does it make sense to call notify() out of locking?

tnqn replied:
I used buffered channel to make it as quick as possible, but yes it can still block if it's not consumed fast enough and less efficient than just moving it out of locking anyway. Then maybe I could just make `notify` execute synchrounouly out of locking to make it simpler and easier to test.

tnqn replied:
Just found we cannot remove locking here as it needs to access i.selectorItems and sItem.groupItemKeys.

jianjuns replied:
What you think about copying the keys and notify after releasing the lock? But if the channel does not really block, or non-blocking does not really help parallelization, might not be worthwhile.

tnqn replied:
If we don't assume the channel can be full with 1000 buffers, it won't block, and copying the keys will introduce some overhead and may use more time than sending items to channels to directly. Maybe let's have a test to see if it can really block or we increate the buffer size later.

jianjuns replied:
Sounds good to me.

jianjuns replied:
Probably good to add some comments before notify() about the possible blocking.

tnqn replied:
Done, and add comment to `AddEventHandler` that eventHandler shouldn't perform blocking operation. I could add time stats to log unexpected long executiont time later.

jianjuns commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,728 @@
+	selectorItems map[string]*selectorItem
+	// selectorItemIndex is nested map from entityType to Namespace to keys of selectorItems.
+	// It's used to filter potential selectorItems when matching an labelItem.
+	selectorItemIndex map[entityType]map[string]sets.String
```
Is it easier to understand if we use a separate map[entityType]sets.String for cluster scope selectors?

jianjuns replied:
Comments definitely help. But I thought your usage pattern is always to check the selector is cluster scope or not, and if so look up with "". Then it does not help simplify the code to reuse a single map? Maybe I missed something.

tnqn replied:
They are treated in same way when maintaing the selectorItemIndex:
https://github.com/vmware-tanzu/antrea/blob/413aad795ae599ef061811a552541f27358b32ed/pkg/controller/grouping/group_entity_index.go#L480
https://github.com/vmware-tanzu/antrea/blob/413aad795ae599ef061811a552541f27358b32ed/pkg/controller/grouping/group_entity_index.go#L505-L509

jianjuns replied:
Got it. Fine with "" (do you think a constant helps).

tnqn replied:
Updated to constant emptyNamespace

tnqn commented:
/test-all

tnqn commented:
/test-all

tnqn commented:
/test-all

tnqn commented:
@jianjuns @antoninbas @Dyanngg I added more unit tests and a method `HasSynced` to avoid intermediate result after restarting since your last review and have addressed all comments if I'm not missing one. Could you take another look?

Dyanngg reviewed (dismissed):
LGTM

antoninbas commented on pkg/controller/grouping/controller_test.go:
```diff
@@ -0,0 +1,102 @@
+			assert.NoError(t, wait.Poll(10*time.Millisecond, time.Second, func() (done bool, err error) {
+				return index.HasSynced(), nil
+			}), "GroupEntityIndex hasn't been synced in 1 second after starting GroupEntityController")
+			assert.Equal(t, len(tt.initialPods), *index.initialPodCount)
+			assert.Equal(t, len(tt.initialPods), index.maxPodCount)
+			assert.Equal(t, len(tt.initialPods), index.curPodCount)
+			assert.Equal(t, len(tt.initialNamespaces), *index.initialNamespaceCount)
+			assert.Equal(t, len(tt.initialNamespaces), index.maxNamespaceCount)
+			assert.Equal(t, len(tt.initialNamespaces), len(index.namespaceLabels))
+			assert.Equal(t, len(tt.initialExternalEntities), *index.initialExternalEntityCount)
+			assert.Equal(t, len(tt.initialExternalEntities), index.maxExternalEntityCount)
+			assert.Equal(t, len(tt.initialExternalEntities), index.curExternalEntityCount)
```
nit: there is an `assert.Len` assertion

tnqn replied:
done

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,818 @@
+	externalEntityType
+)
+
+// GroupType is a public type used to differentiate Groups.
+type GroupType string
```
nit for readability: this type definition would make more sense *before* defining the interface

tnqn replied:
done

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,818 @@
+	}
+}
+
+// deleteGroupFromSelectorItem disconnects an groupItem from a selectorItem.
```
s/an groupItem/a groupItem

tnqn replied:
done

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,818 @@
+				return false
+			}
+			if !sel.NamespaceSelector.Matches(namespaceLabels) {
+				// Pod's Namespace do not match namespaceSelector.
```
s/do not/does not

tnqn replied:
done

antoninbas commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,818 @@
+		}
+		return true
+	}
+	// TODO: if none of the above selector is set, it means matching all entities or none?
+	// This is only possible for ClusterGroup. From previous code, it means matching none.
```
Sorry for resuscitating that conversation, but I feel like the comment should make it clear that "unset" here is either missing (`nil`) or empty. Unless I misunderstood the thread.

It seem that this doesn't need to be a TODO either - this is just the behavior we have and it seems we do not intend to change it.

tnqn replied:
removed TODO and made it more clear.

antoninbas commented on pkg/controller/networkpolicy/networkpolicy_controller_perf_test.go:
```diff
@@ -333,3 +434,75 @@ func BenchmarkSyncAddressGroup(b *testing.B) {
+	benchmarkInit(b, namespaces, networkPolicies, pods)
+}
+
+func BenchmarkInitXLargeScaleWithClusterScopedNetpol(b *testing.B) {
+	namespaces, networkPolicies, pods := getXLargeScaleWithClusterScopedNetpol()
+	benchmarkInit(b, namespaces, networkPolicies, pods)
+}
```
thanks for adding this

antoninbas commented on pkg/controller/networkpolicy/networkpolicy_controller_perf_test.go:
```diff
@@ -333,3 +434,75 @@ func BenchmarkSyncAddressGroup(b *testing.B) {
+	for _, networkPolicy := range networkPolicies {
+		c.addNetworkPolicy(networkPolicy)
+	}
+	for c.appliedToGroupQueue.Len() > 0 {
+		key, _ := c.appliedToGroupQueue.Get()
+		c.syncAppliedToGroup(key.(string))
+		c.appliedToGroupQueue.Done(key)
+	}
+	for c.internalNetworkPolicyQueue.Len() > 0 {
+		key, _ := c.internalNetworkPolicyQueue.Get()
+		c.syncInternalNetworkPolicy(key.(string))
+		c.internalNetworkPolicyQueue.Done(key)
+	}
+	for c.addressGroupQueue.Len() > 0 {
+		key, _ := c.addressGroupQueue.Get()
+		c.syncAddressGroup(key.(string))
+		c.addressGroupQueue.Done(key)
+	}
```
in practice (actual controller), these would be handled by multiple goroutines right?

tnqn replied:
Yes, making the test execute serial is to measure the time and memory accurately when the events are handled in ideal order.

antoninbas reviewed (dismissed):
Thanks for all the work on this. I took another quick look and it looks good to me. I found a couple nits.

tnqn commented:
/test-all

antoninbas reviewed (dismissed):
LGTM

Dyanngg commented:
/test-e2e

jianjuns commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,817 @@
+	// Numbers used to determine whether the GroupEntityIndex has been initialized with the full lists of Pods,
+	// Namespaces, and ExternalEntities.
+	// If a kind's max count is greater than or equal to its initial count, it's considered synced.
+	maxPodCount     int
```
How about accumulatedPodCount?

tnqn replied:
done

jianjuns commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,817 @@
+	// If a kind's max count is greater than or equal to its initial count, it's considered synced.
+	maxPodCount     int
+	curPodCount     int
+	initialPodCount *int
```
Probably add some comments to explain nil means initialPodCount is not initalized yet.
What you think using -1 to indicate initialPodCount is not set?

tnqn replied:
done

jianjuns reviewed (dismissed):
Just a few nits.

tnqn commented:
/test-all

jianjuns commented on pkg/controller/grouping/group_entity_index.go:
```diff
@@ -0,0 +1,827 @@
+
+// notify notifies the affected groups to eventHandlers.
+// It's supposed to be called with the lock held as it accesses the selectorItems. Normally the method shouldn't block
+// as the event channel is buffered and the consumer Run should execute quickly. If it's blocking in practice, we should
```
blocking -> blocked or "it blocks"?

tnqn replied:
done

tnqn commented:
/test-conformance
/test-all-features-conformance
/test-windows-conformance
/test-windows-networkpolicy

tnqn commented:
/test-conformance
/test-e2e

tnqn commented:
/test-networkpolicy
/test-e2e
/test-conformance
